tools/config/__init__.py

```diff
@@ -552,21 +568,42 @@ def _generate_booloader_build(self, target_overrides, rom_start, rom_size):
                 raise ConfigException("bootloader executable does not "
                                       "start at 0x%x" % rom_start)
             part_size = (part.maxaddr() - part.minaddr()) + 1
-            yield Region("bootloader", rom_start + start, part_size, False,
+            start = Config._align_on_sector(rom_start + start, sectors) - rom_start
+            offset = start + rom_start
+            part_size = Config._align_on_sector(offset + part_size, sectors) - offset
+            yield Region("bootloader", offset, part_size, False,
                          filename)
             start += part_size
         if 'target.restrict_size' in target_overrides:
             new_size = int(target_overrides['target.restrict_size'], 0)
-            yield Region("application", rom_start + start, new_size, True, None)
+            start = Config._align_on_sector(rom_start + start, sectors) - rom_start
+            offset = rom_start + start
+            new_size = Config._align_on_sector(offset + new_size, sectors) - offset
+            yield Region("application", offset, new_size, True, None)
             start += new_size
-            yield Region("post_application", rom_start +start, rom_size - start,
+            start = Config._align_on_sector(rom_start + start, sectors) - rom_start
+            yield Region("post_application", rom_start + start, rom_size - start,
                          False, None)
         else:
+            start = Config._align_on_sector(rom_start + start, sectors) - rom_start
             yield Region("application", rom_start + start, rom_size - start,
                          True, None)
         if start > rom_size:
             raise ConfigException("Not enough memory on device to fit all "
                                   "application regions")
+    @staticmethod
+    def _align_on_sector(address, sectors):
+        target_size = -1
+        target_start = -1
+        for (start, size) in sectors:
+            if address < start:
+                break
```

Comment:
If this just rounds up, could it return start here?

Reply:
No. The list of sectors is not a list of all possible sectors. It is a list of tuples with (start of block, size of sectors in that block). Maybe I should have made that more clear.